upf/upf_gtpu_encap.c

```diff
@@ -57,6 +57,238 @@ typedef enum
   UPF_ENCAP_N_NEXT,
 } upf_encap_next_t;
 
+static u16
+encode_error_indication (vlib_buffer_t * b, gtp_error_ind_t * error, int is_ip4)
+{
+  u8 *p = vlib_buffer_get_current (b);
+  u16 length, nlength;
+
+  length = (is_ip4) ? 4 : 16;
+  nlength = clib_host_to_net_u16 (length);
+  /* IE TEID I (5 bytes) + IE GTP Peer address (3 bytes + IPaddr length) + IE Recovery (2 bytes) */
+  b->current_length = 5 + 2 + 3 + length;
+
+  /* Encode TEID I, TEID already in NET order */
+  *p = 16; p++;
+  memcpy (p, &error->teid, sizeof (u32));
+  p += 4;
+  /* Encode Recovery */
+  *p = 14; p++;
+  *p = 0; p++;
+  /* Encode GTP Peer address */
+  *p = 133; p++;
+  memcpy (p, &nlength, sizeof (u16));
+  p += 2;
+  memcpy (p, (length == 4) ? (void *) &error->addr.ip4 : (void *) &error->addr.ip6, length);
+
+  return b->current_length; //bytes encoded
+}
```

Comment:
* no magic constants, please. define the IEs id somewhere 
* no magic lengths for anything, I would define a GTP-U IE struct, with fields for id, length and a variable length array for the content. You can then calculate the length by sizeof(*ie + payload_size).
* GTP extension header for UDP source port is missing, see TS 29.281 (be careful, extension headers are not GTP IEs)

Reply:
Thanks for comments. I will fix constants for more readability and will operate language constructions rather then pure decimal representations.
About GTP extension header UDP Source Port, I did noticed that based on Clause 7.3.1 TS 29.281
`GTP entities may include the "UDP Port" extension header (Type 0x40)`
But I thought it's not mandatory to have in our implementation

Reply:
I've added GTP Extension Header UDP Port encoding.
However, having this present in a packet confuses Scapy a little bit (GTP IE List is now considering as Raw data for the packet). Wireshark pcap looks OK for Error Indication message produced by UPF, so I can assume it's Scapy issue.